radicale/storage.py

```diff
@@ -505,13 +505,38 @@ def update(self, href, vobject_item, etag=None):
     def delete(self, href=None, etag=None):
         # TODO: use etag in code and test it here
         # TODO: use returned object in code
+        def sync_directory(path):
+            if os.name == "posix":
+                fd = os.open(path, 0)
+                try:
+                    if hasattr(fcntl, "F_FULLFSYNC"):
+                        fcntl.fcntl(fd, fcntl.F_FULLFSYNC)
+                    else:
+                        os.fsync(fd)
+                finally:
+                    os.close(fd)
+
         if href is None:
             # Delete the collection
+            tmp_filesystem_path = None
             if os.path.isdir(self._filesystem_path):
-                shutil.rmtree(self._filesystem_path)
+                try:
+                    os.rmdir(self._filesystem_path)
+                except OSError:
+                    while True:
+                        tmp_filesystem_path = os.path.join(
+                            os.path.dirname(self._filesystem_path),
+                            ".Radicale.tmp-" + hex(getrandbits(32))[2:])
+                        if not os.path.exists(tmp_filesystem_path):
+                            break
+                    os.rename(self._filesystem_path, tmp_filesystem_path)
```

Comment:
Couldn't we just use the stdlib tmpdir functions for this? I.e. create a tmpdir, move the collection into that folder and delete the tmpdir. Less performant but also less code to overlook.

Also there'd be a race condition here if there was no global lock.